src/db.js

```diff
@@ -0,0 +1,77 @@
+import Dexie from 'dexie'
+import * as localForage from 'localforage'
+
+const db = new Dexie('EM')
+
+const migrateToDexie = async () => {
+  const THOUGHTINDEX_KEY_START = 'thoughtIndex-'
+  const CONTEXTINDEX_KEY_START = 'contextIndex-'
+  const {
+    cursor,
+    lastUpdated,
+    recentlyEdited,
+    schemaVersion,
+  } = await localForage.getItems([
+    'cursor',
+    'lastUpdated',
+    'recentlyEdited',
+    'schemaVersion',
+  ])
+  const helpersValuesToUpdate = {
+    ...(cursor ? { cursor } : {}),
+    ...(lastUpdated ? { lastUpdated } : {}),
+    ...(recentlyEdited ? { recentlyEdited } : {}),
+    ...(schemaVersion ? { schemaVersion } : {}),
+  }
+  const helperPromise = Object.keys(helpersValuesToUpdate).length ? (async () => db.helpers.put({ id: 'EM', ...helpersValuesToUpdate }))() : {}
+
+  const thoughtIndexes = await localForage.startsWith(THOUGHTINDEX_KEY_START)
+  const thoughtIndexPromises = Object.keys(thoughtIndexes).map(key => {
+    const hash = key.substring(THOUGHTINDEX_KEY_START.length)
+    dbOperations.updateThoughtIndex(hash, thoughtIndexes[key])
+  })
+
+  const contextIndexes = await localForage.startsWith(CONTEXTINDEX_KEY_START)
+  const contextIndexPromises = Object.keys(contextIndexes).map(key => {
+    const hash = key.substring(CONTEXTINDEX_KEY_START.length)
+    dbOperations.updateContextIndex(hash, contextIndexes[key])
+  })
+
+  const migratedItems = await Promise.all([helperPromise, ...thoughtIndexPromises, ...contextIndexPromises])
+  return migratedItems.length
+}
+
+const initHelpers = async () => {
+  const staticHelpersExist = await db.helpers.get({ id: 'EM' })
+  if (!staticHelpersExist) {
+    await db.helpers.add({ id: 'EM' })
+  }
+}
+
+const initDB = async () => {
+  await db.version(1).stores({
+    thoughtIndex: 'id, value, *contexts, created, lastUpdated',
+    contextIndex: 'id, *context, lastUpdated',
+    helpers: 'id, cursor, lastUpdated, recentlyEdited, schemaVersion'
+  })
+  await initHelpers()
+  await migrateToDexie()
+}
+
+export const dbOperations = {
+  updateThoughtIndex: async (id, thought) => db.thoughtIndex.put({ id, ...thought }),
+  deleteThoughtIndex: async id => db.thoughtIndex.delete(id),
+  getThoughtIndexById: async id => db.thoughtIndex.get(id),
+  getThoughtIndexes: async () => db.thoughtIndex.toArray(),
+  updateContextIndex: async (id, context) => db.contextIndex.put({ id, context }),
+  deleteContextIndex: async id => db.contextIndex.delete(id),
+  getContextIndexes: async () => db.contextIndex.toArray(),
+  updateRecentlyEdited: async recentlyEdited => db.helpers.update('EM', { recentlyEdited }),
+  updateSchemaVersion: async schemaVersion => db.helpers.update('EM', { schemaVersion }),
```

Comment:
The difference between put and update is that 'put' behaves like createOrUpdate, whereas update only works if there's a row that exists. 
'update' is used for 'helpers' table because there can never be multiple records in that table. There's only one row with id 'EM'. Put on the other hand is feasible for thoughts and contexts because we don't need to have 2 different functions for create and update.

Reply:
Okay, thanks, that makes sense.